1.16.5 support, FireShots hand swaping, MetalFragments optimisation and dupe fix =)

Dreig-Michihi opened the pull request.

Dreig-Michihi committed:
Added 1.16 support!

Dreig-Michihi committed:
Added FireShots hand swapping. Tap sneak!
Optimized MetalFragments with Iterator.

Dreig-Michihi committed:
Fixed MetalFragments hoppers dupe!
It may also fix other random dupes with items used for abilities rendering!

Dreig-Michihi committed:
FireShots's default hand depends on player's main hand.

Dreig-Michihi committed:
Removed flame effect when dupe prevented

Dreig-Michihi committed:
MetalFragments added ITEM_CRACK particles

Dreig-Michihi committed:
Added configurable pointer for bending board.
Why: Idk why, but 1-symbol pointers rendered buggy in bending board, when ability's element is subelement and abil is on cooldown. Dound with metal, sand abilities.'
Pointer "> " looks fine.
Also string ">[Color][StrikeThrough]AbilityName" looks normally in chat, but it becomes white in scoreboard!
If pointer has 2 symbols, then everything is okay.

Dreig-Michihi commented:
So?

Dreig-Michihi committed:
FirePunch stockpiling fix!
Very important bug fix.

Dreig-Michihi committed:
Fixed freezes caused by AirBreath on latest minecraft versions.
Bug was found on 1.19.2
Actually in this commit I just refactored JCMethods#extinguishBlocks method!

Aztlon commented on src/com/jedk1/jedcore/JCMethods.java:
```diff
@@ -17,27 +23,34 @@
-import com.projectkorra.projectkorra.ability.CoreAbility;
-import com.projectkorra.projectkorra.ability.util.ComboManager;
-import com.projectkorra.projectkorra.util.TempBlock;
+import java.util.*;
```
this doesn't matter much but next time can you not rearrange the imports

Dreig-Michihi replied:
Yes, I just didn't knew I had default setting in IntelliJ IDEA at that moment. Now I set that imports will never be changed like that.

Aztlon commented on src/com/jedk1/jedcore/ability/earthbending/MetalFragments.java:
```diff
@@ -246,20 +242,21 @@ public void progress() {
-					f.remove();
 				}
 			}
+			if(touchedLiving||f.isOnGround()||f.isDead()) {
```
can you conform to the standard style a little more in the future? with spaces between the parts of this if statement

Dreig-Michihi replied:
Here I was scared, that I will ruin imports again with Alt+Ctrl+L :D 
Now I know, that I can use reformatting on only highlighted part of code.

Aztlon commented on src/com/jedk1/jedcore/ability/earthbending/MetalFragments.java:
```diff
@@ -289,6 +286,7 @@ public void dropSources() {
 
 	public void removeFragments() {
 		for (Item i : thrownFragments) {
+			ParticleEffect.ITEM_CRACK.display(i.getLocation(),3,0.3,0.3,0.3, 0.2, i.getItemStack());
```
same here, more spaces

Dreig-Michihi replied:
Okay!

Aztlon commented on src/com/jedk1/jedcore/ability/firebending/FireShots.java:
```diff
@@ -28,369 +28,388 @@
-	public boolean isEnabled() {
-		ConfigurationSection config = JedCoreConfig.getConfig(this.player);
-		return config.getBoolean("Abilities.Fire.FireShots.Enabled");
-	}
```
let's keep the project in tabs. no need to convert to spaces

Aztlon commented on src/com/jedk1/jedcore/ability/waterbending/Drain.java:
```diff
@@ -25,29 +23,15 @@
-import java.util.HashMap;
-import java.util.List;
-import java.util.Random;
+import java.util.*;
```
why?

Dreig-Michihi replied:
as I wrote before, had default settings in IntelliJ IDEA at that moment.

Aztlon commented on src/com/jedk1/jedcore/ability/firebending/FireShots.java:
```diff
@@ -28,369 +28,388 @@
+    @Attribute("CollisionRadius")
+    private double collisionRadius;
+
+    private Boolean flameInMainHand = null;
```
why does this need to be null by default? why not just true?

Dreig-Michihi replied:
Because hand swapping should be ignored at the first time, when player enabled the ability (when player sneaks up after sneak down).
FireShots enabled with sneak_down, and after that he'll sneak up for sure. First time his hand should not be swapped.

Aztlon replied:
@Dreig-Michihi

Dreig-Michihi replied:
> @Dreig-Michihi

?

Aztlon commented on src/com/jedk1/jedcore/listener/JCListener.java:
```diff
@@ -291,4 +293,11 @@ public void onChat(AsyncPlayerChatEvent event){
+	@EventHandler(priority = EventPriority.NORMAL)
+	public void onItemPickup(InventoryPickupItemEvent event){
+		if(event.getItem().getPickupDelay()>=Short.MAX_VALUE) {
+			event.setCancelled(true);
```
i think i know what this does but can you explain your reasoning for adding this?

Dreig-Michihi replied:
It fixes the bug with duping MetalFragments's ingots. If someone use the same technique for rendering abilities (items with max pickup delay), then their abils dupe will be fixed too.

Aztlon replied:
@Dreig-Michihi

Dreig-Michihi replied:
> @Dreig-Michihi

?

Aztlon reviewed (commented):
I plan to approve this, but I have a few comments and questions

Dreig-Michihi committed:
Turned imports back how they were.

FireShots should be clean now.

AbilityListener now listens to ability's class instead of just name. (Except WaterManipulation used in combos)

Combustion's charging particles circle now starts from the point where player looks.

Fixed bug when Combustion couldn't start when previous combustion's blocks weren't reverted yet.

Fixed cooldowns. Now Combustion will be set on cooldown right after launch or being damaged while charging.

Dreig-Michihi commented:
@Aztlon can you check the latest commit?

Dreig-Michihi committed:
FireShots pain

Dreig-Michihi committed:
FireShots pain

Dreig-Michihi committed:
Combustion bug fix

Dreig-Michihi reviewed (commented):
Submit

StrangeOne101 commented on pom.xml:
```diff
@@ -4,7 +4,7 @@
   <groupId>com.jedk1</groupId>
   <artifactId>jedcore</artifactId>
-  <version>2.12.1-Spigot1.18-PK1.10.2</version>
+  <version>2.12.2-Spigot1.16-PK1.10.2-BETA</version>
```
Maybe remove BETA from the version?
